### PR TITLE
feat: refactor and untangle analytics

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

This PR is a biiiiig refactor on analytics. 

There was many issues with analytics. The main motivation is that cow.fi analytics was broken, but when I tried to inspect on the issue I saw a big mess that was a bit time-consuming to untangle. 

I added some comments in the code.  It refactors the 4 apps so they all use the same analytics abstractions and logic.
Reorganise all analytics code.

I added some change that is unrelated and might be lost in so many changes (pls forgive me), this is why I point it out here: https://github.com/cowprotocol/cowswap/pull/4718/files#r1678381293 

# Test
Just make sure I didn't break anything in the 4 apps with this PR.  cow.fi issues should be gone, so now we should see the explorer event being triggered.
I tried to keep the current logic.

Make sure you enable the analytics debugging during your tests.
image

I refactor all External Links in the apps, so it's possible there's some difference in some links.

### CODE REVIEW

File: apps/cow-fi/data/cow-amm/const.tsx
Comment (anxolin) [inline apps/cow-fi/data/cow-amm/const.tsx:182]: In general, all event triggering has been to a `events.ts` file per app that has all analytics events and abstract about the analytics implementation details

File: apps/cow-fi/modules/analytics/events.ts
Comment (anxolin) [inline apps/cow-fi/modules/analytics/events.ts:5]: Now, there's one enum `Category` per app. Before apps were sharing the same categories and there was some categories not part of the enum

**Comment by fairlighteth** [comment]
The explorer still has the `test` ID on:
Screenshot 2024-07-16 at 17 55 51
Screenshot 2024-07-16 at 17 56 21

**Comment by fairlighteth** [comment]
Otherwise, Swap, Widget and Cow.fi seemed to load the correct tag ID

File: apps/explorer/tsconfig.json
Comment (alfetopito) [inline apps/explorer/tsconfig.json:24]: Should we import the wallet into the Explorer?
Probably shouldn't be needed for analytics.

File: libs/analytics/src/googleAnalytics/initGoogleAnalytics.ts
Comment (alfetopito) [inline libs/analytics/src/googleAnalytics/initGoogleAnalytics.ts:12]: Newbie question: does this mean that SSR won't have client ids?
Of course not, but the client side won't ever be hydrated with this sort of info?

File: libs/analytics/src/googleAnalytics/initGoogleAnalytics.ts
Comment (alfetopito) [inline libs/analytics/src/googleAnalytics/initGoogleAnalytics.ts:15]: I take we look for both to accept both app types. Is there ever a chance this might cause a conflict?
Would it work if we set a single one for both?

File: libs/analytics/src/pixels/initPixelAnalytics.ts
Comment (alfetopito) [inline libs/analytics/src/pixels/initPixelAnalytics.ts:24]: Does this mean all of those are on by default on prod now?

File: libs/ui/src/pure/ExternalLink/index.tsx
Comment (alfetopito) [inline libs/ui/src/pure/ExternalLink/index.tsx:124]: I know, I know, but this is a pure component.
Should not have side effects/hooks.

File: libs/ui/src/pure/Footer/index.tsx
Comment (alfetopito) [inline libs/ui/src/pure/Footer/index.tsx:292]: Another pure component 🙈 

Maybe best move them to containers or something in case there's no easy solution?

**Review by alfetopito — COMMENTED** [review]
Some comments, have not tested

File: libs/ui/src/pure/Footer/index.tsx
Comment (anxolin) [reply libs/ui/src/pure/Footer/index.tsx:292]: Yes, good point will do

**Comment by elena-zh** [comment]
Hey @anxolin , great job!
Questions/issues:
1. Is this OK that I see this error on console when open Explorer app?
image
2. Explorer: I don't see 'Fired' event con console when press on the address link/TX hash for a filled order
image
3. Cow.fi (not sure): these links navigation do not have an analytics event recorded
image
image
4. Get protected button does not have an event on console  (same on Prod, btw, so not an issue related to the current PR): 
image
5. on CoW Swap connect/disconnect/change wallet events are recorded with no details: 
image
6. CoW Swap: no events recorded for links in the header (same on Prod, btw, so not an issue related to the current PR)
image
image


Thanks!

File: libs/analytics/src/googleAnalytics/initGoogleAnalytics.ts
Comment (anxolin) [reply libs/analytics/src/googleAnalytics/initGoogleAnalytics.ts:12]: I think it will, at least this is what's happening right now. I saw it was working with this approach. It won't try to persist anything in the local storage for SSR rendering but it will for client rendering. I didn't get too deep on why it was working... as it just works :)

File: libs/analytics/src/googleAnalytics/initGoogleAnalytics.ts
Comment (anxolin) [reply libs/analytics/src/googleAnalytics/initGoogleAnalytics.ts:15]: > I take we look for both to accept both app types. Is there ever a chance this might cause a conflict?

No, unless you define both. In fact they need the same ID. What I can do is to make this a parameter of the init, so each app decides. I just left it like this to avoid a bit repetition. Happy to move it out to the apps if you prefer

File: libs/analytics/src/pixels/initPixelAnalytics.ts
Comment (anxolin) [reply libs/analytics/src/pixels/initPixelAnalytics.ts:24]: Yes, good point.

how it works now is that it relies on the factory method to decide if this is needed or not.

```
if (!isPixelEnabled) {
    return undefined
  }
```

And this is currently using the same logic as before:
https://github.com/cowprotocol/cowswap/blob/e2fbecba5ec1c0b49316643d8ddf1fb443b187b3/libs/analytics/src/pixel/constants.ts#L3

replaced with:

```ts
const isPixelEnabled = isProd || isEns
```

However, its a good point, because before it was enable on prod, but it had no effect because we commented out the pixels https://github.com/cowprotocol/cowswap/blob/e2fbecba5ec1c0b49316643d8ddf1fb443b187b3/apps/cowswap-frontend/src/common/hooks/useAnalyticsReporter.ts#L104

What I will do is to make it return undefined always :P

Comment (anxolin) [reply libs/analytics/src/pixels/initPixelAnalytics.ts:24]: 6b75e6f37e5b0eda69c2cff6a5805dc29027839d

File: libs/ui/src/pure/ExternalLink/index.tsx
Comment (anxolin) [reply libs/ui/src/pure/ExternalLink/index.tsx:124]: 599906b707d5d0b8e9ca845b8c3069ccd3c8baa5

File: libs/ui/src/pure/Footer/index.tsx
Comment (anxolin) [reply libs/ui/src/pure/Footer/index.tsx:292]: 599906b707d5d0b8e9ca845b8c3069ccd3c8baa5

**Comment by fairlighteth** [comment]
Looks good now 👍🏼

**Comment by anxolin** [comment]
@elena-zh 

1. Regarding the test issue, should be fixed now in the PR preview. We still show this error in localhost development, but this is nothing new. The code is equivalent as before. I don't think it breaks anything as it allows us to debug analytics events and this environment is for test porpoises only

This screenshot is from develop, where we already have this warning:
image



2. I think the issue is the outbound links are not relevant anymore. This PR I tried to keep the same logic, but as you can see here Google Analytics 4 deprecated this function. I checked and we also don't trigger it ourselves in production (so this PR maintains same logic). The outbound links are automatically tracked by analytics, so in principle should be fine to remove all this (but didn't want to introduce new changes now to not overcomplicate the PR, scope went crazy already)

image

3. I do see a `page_view` event. Maybe not the click on the menu. But I don't think we had this before this PR. This PR is not about adding missing events, we could potentially add a lot more I agree

4. Same as above

5. Great catch! 🚨 I will review what's wrong with this

6. Same as 3 and 4. You can suggest to track more things (if someone plans to look at that data and we think is impactful). For now removing from the scope of this PR

**Comment by anxolin** [comment]
@elena-zh solved!

Great catch with (5)

image


Also, cool to check the dimensions of the events:

image

File: apps/explorer/tsconfig.json
Comment (alfetopito) [reply apps/explorer/tsconfig.json:24]: Yes, I rather remove it please

**Review by alfetopito — COMMENTED** [review]
Approved!

Leaving the real approval to Elena as once it has an approval it'll auto-merge.

Also, check the linter.

**Review by elena-zh — APPROVED** [review]
Great, thank you!

**Pull request merged**
